src/iterative_ensemble_smoother/ies.py

```diff
@@ -0,0 +1,385 @@
+
+    relative_energy = sigma / total_sigma
+    return np.searchsorted(relative_energy, truncation, side="left") + 1
+
+
+def truncated_svd_inv_sigma(S, truncation, full_matrices=False):
+    """
+    Compute truncated SVD of a matrix S, keeping a fraction `truncation` of the
+    total energy (singular values squared).
+
+    Examples
+    --------
+    >>> S = np.diag([100, 1, 0]) #  Create a singular matrix
+    >>> truncated_svd_inv_sigma(S, 1.0)
+    (array([[1., 0.],
+           [0., 1.],
+           [0., 0.]]), array([0.01, 1.  ]), array([[1., 0., 0.],
+           [0., 1., 0.]]))
+
+    Keep only 90 % of energy, removing one non-zero singular value:
+
+    >>> S = np.diag([100, 1, 0])
+    >>> truncated_svd_inv_sigma(S, 0.9)
+    (array([[1.],
+           [0.],
+           [0.]]), array([0.01]), array([[1., 0., 0.]]))
+
+    """
+
+    U, singular_values, VT = sp.linalg.svd(
+        S,
+        full_matrices=full_matrices,
+        compute_uv=True,
+        overwrite_a=False,
+        check_finite=True,
+        lapack_driver="gesdd",
+    )
+
+    if isinstance(truncation, float):
+        num_significant = calc_num_significant(singular_values, truncation)
+    else:
+        num_significant = truncation
+
+    assert np.all(
+        singular_values[:num_significant] > 0
+    ), "Must have positive singular values"
+
+    inverted_singular_values = np.zeros_like(singular_values)
+    inverted_singular_values[:num_significant] = 1 / singular_values[:num_significant]
+
+    return (
+        U[:, :num_significant],
+        inverted_singular_values[:num_significant],
+        VT[:num_significant, :],
+    )
+
+
+def exact_inversion(W, S, H, steplength):
+    """Compute exact inversion, assuming identity error covariance matrix.
+
+    This is equation (51) in the paper.
+
+    W has shape (num_ensemble, num_ensemble)
+    S has shape (num_outputs, num_ensemble)
+    H has shape (num_outputs, num_ensemble)
+
+    Returns
+    -------
+    W : np.ndarray
+        Updated matrix W_i+1.
+
+    Examples
+    --------
+    >>> W = np.array([[1, 2], [3, 4]])
+    >>> S = np.array([[1, 2], [3, 4], [5, 6]])
+    >>> H = np.array([[0, 1], [1, 0], [1, 1]])
+
+    Naive computation:
+
+    >>> W - 1 * (W - np.linalg.inv(S.T @ S + np.eye(2)) @ S.T @ H)
+    array([[ 0.13793103, -0.0862069 ],
+           [ 0.06896552,  0.20689655]])
+
+    This function:
+
+    >>> exact_inversion(W, S, H, 1.0)
+    array([[ 0.13793103, -0.0862069 ],
+           [ 0.06896552,  0.20689655]])
+
+    """
+    # Section 3.2 - Exact inversion assuming diagonal error covariance matrix
+
+    # Equation (51) states that
+    # W_{i+1} = W_i - step * (W_i - (S.T @ S + I)^-1 @ S.T @ H)
+    # Here we form
+    # C = S.T @ S + I
+    # This is a square symmetric matrix, instead of taking the SVD we
+    # call sp.linalg.solve. This is equally fast.
+
+    # Form the matrix C
+    _, ensemble_size = S.shape
+    C = S.T @ S  # TODO: Only form upper part of this matrix
```

Comment:
It is done in the same way in the C++ code, so okay in this PR. The todo should be create issue to only form upper part of this matrix.